linuxkm/lkcapi_rsa_glue.c

```diff
@@ -0,0 +1,1830 @@
+        #endif /* WOLFKM_DEBUG_RSA */
+        err = -EINVAL;
+        goto pkcs1_dec_out;
+    }
+
+    if (dec_len > req->dst_len) {
+        err = -EOVERFLOW;
+        goto pkcs1_dec_out;
+    }
+
+    /* copy dec to req->dst */
+    scatterwalk_map_and_copy(dec, req->dst, 0, dec_len, 1);
+
+    err = 0;
+pkcs1_dec_out:
+    if (enc != NULL) { free(enc); enc = NULL; }
+    if (dec != NULL) { free(dec); dec = NULL; }
+
+    #ifdef WOLFKM_DEBUG_RSA
+    pr_info("info: exiting km_pkcs1_dec\n");
+    #endif /* WOLFKM_DEBUG_RSA */
+    return err;
+}
+
+#if defined(LINUXKM_DIRECT_RSA) && defined(WC_RSA_NO_PADDING)
+/**
+ * Tests implemented below.
+ * */
+static int linuxkm_test_rsa(void)
+{
+    int rc = 0;
+    rc = rsa_no_pad_test();
+    if (rc != 0) {
+        pr_err("rsa_no_pad_test() failed with retval %d.\n", rc);
+        return rc;
+    }
+
+    #ifdef WOLFSSL_KEY_GEN
+    /* test wolfcrypt RSA API vs wolfkm RSA driver. */
+    rc = linuxkm_test_rsa_driver(WOLFKM_RSA_DRIVER, 2048);
+    if (rc) { return rc; }
+
+    #ifdef WOLFKM_DEBUG_RSA
+    rc = linuxkm_test_rsa_driver(WOLFKM_RSA_DRIVER, 3072);
+    if (rc) { return rc; }
+
+    rc = linuxkm_test_rsa_driver(WOLFKM_RSA_DRIVER, 4096);
+    if (rc) { return rc; }
+
+
+    /* repeat test against stock linux RSA akcipher. */
+    rc = linuxkm_test_rsa_driver("rsa-generic", 2048);
+    if (rc) { return rc; }
+
+    rc = linuxkm_test_rsa_driver("rsa-generic", 3072);
+    if (rc) { return rc; }
+
+    rc = linuxkm_test_rsa_driver("rsa-generic", 4096);
+    if (rc) { return rc; }
+    #endif /* WOLFKM_DEBUG_RSA */
+    #endif /* WOLFSSL_KEY_GEN */
+
+    return rc;
+}
+#endif /* LINUXKM_DIRECT_RSA */
+
+#ifndef NO_SHA256
+static int linuxkm_test_pkcs1_sha256(void)
+{
+    int rc = 0;
+
+    #ifdef WOLFSSL_KEY_GEN
+    rc = linuxkm_test_pkcs1_driver(WOLFKM_PKCS1_SHA256_DRIVER, 2048,
+                                   SHA256h, 32);
+    if (rc) { return rc; }
+
+    #ifdef WOLFKM_DEBUG_RSA
+    rc = linuxkm_test_pkcs1_driver(WOLFKM_PKCS1_SHA256_DRIVER, 3072,
+                                   SHA256h, 32);
+    if (rc) { return rc; }
+
+    rc = linuxkm_test_pkcs1_driver(WOLFKM_PKCS1_SHA256_DRIVER, 4096,
+                                   SHA256h, 32);
+    if (rc) { return rc; }
+
+    /* repeat test against stock linux pkcs1pad. */
+    rc = linuxkm_test_pkcs1_driver("pkcs1pad(rsa-generic,sha256)", 2048,
+                                   SHA256h, 32);
+    if (rc) { return rc; }
+
+    rc = linuxkm_test_pkcs1_driver("pkcs1pad(rsa-generic,sha256)", 3072,
+                                   SHA256h, 32);
+    if (rc) { return rc; }
+
+    rc = linuxkm_test_pkcs1_driver("pkcs1pad(rsa-generic,sha256)", 4096,
+                                   SHA256h, 32);
+    if (rc) { return rc; }
+    #endif /* WOLFKM_DEBUG_RSA */
+    #endif /* WOLFSSL_KEY_GEN */
+
+    return rc;
+}
+#endif /* !NO_SHA256 */
+
+#ifdef WOLFSSL_SHA512
+static int linuxkm_test_pkcs1_sha512(void)
```

Comment:
This is kind of overzealous testing, but the kernel `crypto/testmgr.c` code seems to skip many pkcs1pad variations.